src/MessageHandler/DatasetSubmissionFilerHandler.php

```diff
@@ -0,0 +1,157 @@
+     */
+    private $logger;
+
+    /**
+     * Instance of symfony messenger message bus.
+     *
+     * @var MessageBusInterface
+     */
+    private $messageBus;
+
+    /**
+     * The entity event dispatcher.
+     *
+     * @var EntityEventDispatcher
+     */
+    protected $entityEventDispatcher;
+
+    /**
+     * The entity manager.
+     *
+     * @var EntityManagerInterface
+     */
+    protected $entityManager;
+
+    /**
+     * DatasetSubmissionFilerHandler constructor.
+     *
+     * @param DatasetSubmissionRepository $datasetSubmissionRepository Dataset Submission Repository.
+     * @param LoggerInterface             $filerLogger                 Name hinted filer logger.
+     * @param MessageBusInterface         $messageBus                  Symfony messenger bus interface instance.
+     * @param EntityEventDispatcher       $entityEventDispatcher       The entity event dispatcher.
+     * @param EntityManagerInterface      $entityManager               The entity manager.
+     */
+    public function __construct(
+        DatasetSubmissionRepository $datasetSubmissionRepository,
+        LoggerInterface $filerLogger,
+        MessageBusInterface $messageBus,
+        EntityEventDispatcher $entityEventDispatcher,
+        EntityManagerInterface $entityManager
+    ) {
+        $this->datasetSubmissionRepository = $datasetSubmissionRepository;
+        $this->logger = $filerLogger;
+        $this->messageBus = $messageBus;
+        $this->entityEventDispatcher = $entityEventDispatcher;
+        $this->entityManager = $entityManager;
+    }
+
+    /**
+     * Invoke function to process dataset submission filer.
+     *
+     * @param DatasetSubmissionFiler $datasetSubmissionFiler Dataset submission filer message to be handled.
+     */
+    public function __invoke(DatasetSubmissionFiler $datasetSubmissionFiler)
+    {
+        $datasetSubmissionId = $datasetSubmissionFiler->getSubmissionId();
+        $datasetSubmission = $this->datasetSubmissionRepository->find($datasetSubmissionId);
+        $dataset = $datasetSubmission->getDataset();
+        $loggingContext = array(
+            'dataset_id' => $dataset->getId(),
+            'udi' => $datasetSubmission->getDataset()->getUdi(),
+            'dataset_submission_id' => $datasetSubmissionId
+        );
+        $fileset = $datasetSubmission->getFileset();
+
+        if ($fileset instanceof Fileset) {
+            // Log processing complete.
+            $this->logger->info('Dataset submission process started', $loggingContext);
+            foreach ($fileset->getFiles() as $file) {
+                $this->processFile($file, $loggingContext);
+            }
+
+            $datasetSubmission->setDatasetFileTransferStatus(
+                DatasetSubmission::TRANSFER_STATUS_COMPLETED
+            );
+            $dataset->updateAvailabilityStatus();
+            $this->entityManager->persist($datasetSubmission);
+            $this->entityManager->persist($dataset);
```

Comment:
Do these really have to be persisted? Doesn't that mean they have recreated?

Reply:
Yes, because we update the availability status of the dataset

Reply:
I think you only persist when you want to start tracking an object, this object is already being tracked. A flush should do. I'm have a bit of a hard time testing this.

Reply:
I'm not sure how it is already being tracked?